### ISSUE TITLE
jsondecode function

### ISSUE THREAD

**Issue by matti** [hashicorp/terraform#10363]
jsonencode could be used as a way to pass lists and maps to module if there was a jsondecode function (or jsonparse etc).

**Comment by ckyoog** [comment]
And better support **nested** map and list.

**Comment by apparentlymart** [comment]
Hi @matti and @ckyoog! This is an old issue, but I have some news about it!

The current configuration language can't support a function like this because it has no way to model a function whose return type depends on its input value.

For a little while now we've been working on a revamp of the configuration language to address a number of shortcomings and limitations, and this is one of the things the new implementation can now support.

We're going to be rolling this new implementation out cautiously because it's a big change that will almost definitely contain some small breaking changes that we need to manage. However, we are planning to release an opt-in preview version of it soon to gather feedback and find as many bugs as possible. This initial release should include the new `jsondecode` function, though it will still be subject to some limitations initially since it will take a few steps to update all of Terraform's internals to support the new system.

**Comment by ckyoog** [comment]
Thank you @apparentlymart . It's really a good news. I can't wait to see what brilliant features the new implementation can bring to us.

**Comment by matti** [comment]
just found that @ewbankkit has done this: https://github.com/EvilSuperstars/terraform-provider-jsondecode

**Comment by matti** [comment]
"The subset of JSON that can be decoded is limited - boolean, number, string, object with string values and array of strings." - see tests: https://github.com/EvilSuperstars/terraform-provider-jsondecode/blob/master/decode/data_source_decode_test.go

**Comment by matti** [comment]
I also gave a try, not as good, but might help someone https://github.com/matti/terraform-json-map

**Comment by robinbowes** [comment]
Is `jsondecode` due to be released in 0.12 ?

**Comment by ewbankkit** [comment]
I'm hoping that _Reliable JSON syntax_ promised [here](https://www.hashicorp.com/blog/terraform-0-1-2-preview) will give us what we need.

**Comment by robinbowes** [comment]
That's not the same as being able to process json.

**Comment by apparentlymart** [comment]
Although it's not included in the list in the article there, `jsondecode` is planned to be in the 0.12 release, along with improvements to the `jsonencode` function so it can work with arbitrarily-nested types rather than just with lists and maps of strings.

@robinbowes is correct that this is something separate than "Reliable JSON syntax", which is about Terraform's handling of its own configuration files written in JSON format (`.tf.json` files, in other words).

**Comment by robinbowes** [comment]
Do you have any idea when 0.12 will be released? Is it says/weeks/months away?

**Comment by apparentlymart** [comment]
There are more details on the 0.12 release in [the announcement blog post](https://www.hashicorp.com/blog/terraform-0-1-2-preview).

**Comment by apparentlymart** [comment]
I just tried the following in v0.12.0-alpha1:

```hcl
output "foo" {
  value = jsondecode(<<EOT
{
  "foo": "bar",
  "baz": ["boop"]
}
EOT
  )
}
```

```
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

foo = {
  "baz" = [
    "boop",
  ]
  "foo" = "bar"
}
```

The `jsondecode` function is available for experimental use in the v0.12.0-alpha1 release and will be included in the eventual v0.12.0 final release.

Thanks for filing this feature request, and for the patience while we got through the groundwork to make a function like this possible.

**Comment by udondan** [comment]
Already commented on #4789 but also related here:

Here's a partial solution for Terraform v0.11 without any dependencies:

```
variable json_input {
  default = <<EOT
{
  "foo": "bar",
  "baz": "boop"
}
EOT
}

data "external" "helper" {
  program = ["echo", "${var.json_input}"]
}
```

`data.external.helper.result` is the full map stored in `secret_string` and elements can be accessed as `data.external.helper.result["example"]`

Or throw it into a module and return the whole map as output:

```
output "parsed" {
  value = "${data.external.helper.result}"
}
```

```
$ terraform apply
data.external.helper: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

parsed = {
  baz = boop
  foo = bar
}
```

The only limitation is, that the [external data provider can apparently not deal with arrays](https://github.com/terraform-providers/terraform-provider-external/issues/2). So only works with json maps...

> json: cannot unmarshal array into Go value of type string
